### PR TITLE
ENH: Refactor Custom UI classes to use ServiceOfferExtensionPlugin

### PR DESCRIPTION
**Pull request by flongford**
This PR requires force-h2020/force-bdss#228

It generally refactors out the `UIExtensionPlugin` class that was introduced into `force-wfmanager` in #324 and replaces it with `ServiceOfferExtensionPlugin`, imported from `force-bdss`. Doing so provides a generic method for developers to contribute both custom-made `DataViews` and `ContributedUI` objects to `force-wfmanager` via Envisage's `ServiceOffer` protocol. 

An example of how this new functionality can be used will be provided in an upcoming PR force-h2020/force-bdss-plugin-enthought-example#28.

Change Log
--------------
- `UIExtensionPlugin ` subclass replaced by `ServiceOfferExtensionPlugin`
 - Very basic traits `Interface` subclasses `IDataView` and `IBasePlot` introduced, which are designed to be used by `ServiceOfferExtensionPlugin`. These could be expanded upon in order to provide further type / method implementation checking for contributed data views.
- Docstrings for `IContributedUI`, `IDataView` and `IBasePlot` have been updated and direct to force-bdss documentation for information regarding their implementation.

Other changes
--------------
~~A couple of minor changes have also been performed to clarify the differences between `BasePlot` subclasses, coinciding with the introduction of `IBasePlot`~~
~~- `Plot` subclass has been renamed to `ColorPlot` to better represent its functionalities~~
~~- `TestAnyPlot` class has been renamed to `PlotBaseTestCase` and moved to a new separate file.~~
~~- Pyface `GuiTestAssistant` implemented in `TestColorPlot` to perform unit testing of `edit_traits` call~~
NOTE - All these changes can now be found in [enh/color-plot](https://github.com/force-h2020/force-wfmanager/tree/enh/color-plot)

### CODE REVIEW

File: force_wfmanager/ui/review/i_base_plot.py [hunk 1]
```diff
@@ -0,0 +1,20 @@
+from chaco.api import ArrayDataSource
+from traits.api import Instance
+
+from .i_data_view import IDataView
+
+
+class IBasePlot(IDataView):
+    """Envisage required interface for the BasePlot class.
```
Comment (corranwebster) [inline force_wfmanager/ui/review/i_base_plot.py]: These comments are a bit misleading - it isn't required by Envisage.  Interfaces like this are more of an opportunity for things that don't inherit from `BasePlot` to still participate/be contributed/etc. as-if they do.

So more neutral wording like "Interface definition for plots."

Also, should this be `IPlot`, not `IBasePlot`?

File: force_wfmanager/ui/review/i_data_view.py [hunk 2]
```diff
@@ -0,0 +1,14 @@
+from traits.api import Interface, Instance
+
+from force_wfmanager.model.analysis_model import AnalysisModel
+
+
+class IDataView(Interface):
+    """Envisage required interface for the BaseDataView class.
```
Comment (corranwebster) [inline force_wfmanager/ui/review/i_data_view.py]: Same comment here.

File: force_wfmanager/ui/contributed_ui/i_contributed_ui.py [hunk 3]
```diff
@@ -3,6 +3,11 @@
 
 
 class IContributedUI(Interface):
+    """Envisage required interface for the ContributedUI class.
```
Comment (corranwebster) [inline force_wfmanager/ui/contributed_ui/i_contributed_ui.py]: Same comment here.

File: force_wfmanager/ui/review/i_base_plot.py (hunk 1 above)
Comment (flongford) [reply force_wfmanager/ui/review/i_base_plot.py]: > These comments are a bit misleading - it isn't required by Envisage. Interfaces like this are more of an opportunity for things that don't inherit from `BasePlot` to still participate/be contributed/etc. as-if they do.
> 
> So more neutral wording like "Interface definition for plots."

Agreed. The docstrings of all the `Interface` classes introduced have been amended as suggested. They also point to the force-bdss documentation for instructions as to how to implement them.

> Also, should this be `IPlot`, not `IBasePlot`?

The `IBasePlot` subclass was originally intended as an interface around the generic 2D scatter plot defined by `BasePlot` - the `Plot` class just adds an extra option for colouring data points by another variable. However, this naming seems a little confusing, so I've now renamed the `Plot` class to `ColorPlot` and included it in a separate file `color_plot.py`.

File: force_wfmanager/ui/review/tests/test_plot.py [hunk 4]
```diff
@@ -1,291 +1,38 @@
...
-from force_wfmanager.model.analysis_model import AnalysisModel
-from force_wfmanager.ui.review.plot import BasePlot, Plot
+from force_wfmanager.ui.review.base_plot import BasePlot
+from force_wfmanager.ui.review.color_plot import ColorPlot
 from traits.api import push_exception_handler, TraitError
 
-push_exception_handler(reraise_exceptions=True)
-
-
-class TestAnyPlot(object):
```
Comment (flongford) [inline force_wfmanager/ui/review/tests/test_plot.py]: This class provides unit test methods for `TestBasePlot` and `TestColorPlot` classes - I've moved it into a separate file and renamed it `BasePlotTestCase`.

File: force_wfmanager/ui/review/tests/plot_base_test_case.py [hunk 5]
```diff
@@ -0,0 +1,267 @@
...
+from chaco.api import Plot as ChacoPlot
+from chaco.api import BaseXYPlot
+
+from force_wfmanager.model.analysis_model import AnalysisModel
+from traits.api import push_exception_handler
+
+push_exception_handler(reraise_exceptions=True)
+
+
+class PlotBaseTestCase:
```
Comment (flongford) [inline force_wfmanager/ui/review/tests/plot_base_test_case.py]: This class has just been lifted from `TestAnyPlot` and renamed.

**Comment by flongford** [comment]
> Some minor comments about docstrings, but otherwise looks OK.
> 
> I'd encourage using full documentation of the interface, even if it duplicates what is in the base class. In some sense it is the interface which is the source of truth about what is expected.

Amendments have been made to docstrings, taking on board comments and improving documentation of each subclass. Latest commits have also refactored some unit tests in keeping with the renaming of `Plot` to `ColorPlot`. This, combined with a couple of renamed files, seems to have significantly increased the diff and probably contributing to a lower codecov diff coverage.

**Comment by flongford** [comment]
> ..Latest commits have also refactored some unit tests in keeping with the renaming of `Plot` to `ColorPlot`. This, combined with a couple of renamed files, seems to have significantly increased the diff.

Commits relating to this refactoring have now been reverted, with the changes retained in a new branch: `enh/color-plot`. It is expected that this new branch can also be used to address #310

File: force_wfmanager/ui/review/i_base_plot.py (hunk 1 above)
Comment (flongford) [reply force_wfmanager/ui/review/i_base_plot.py]: Update: `Plot` --> `ColorPlot` refactoring has been lifted to a new branch: `enh/color-plot`

**Pull request merged**